Refactor to functional programming

ThijsTyZ opened the pull request:
Refactor object `DisposableEventListener` to function `addEventListener`. This functional approach is simpler and easier to manage.

Bump major version number since this is a breaking change.

mediamonks-thijs committed:
Refactor DisposableEventListener so it has no dependencies

mediamonks-thijs committed:
Remove declaration of members from the properties

mediamonks-thijs committed:
Add unit test

mediamonks-thijs committed:
Add unit test for testing compatibility with seng-event

mediamonks-thijs committed:
move @babel/runtime back to the dependencies list

mediamonks-thijs committed:
upgrade typescript to latest version in order to use Parameters<>

mediamonks-thijs committed:
Update typings to align with Element.addEventListener and Element.removeEventListener

mediamonks-thijs committed:
Update tsconfig.json for include and exclude

mediamonks-thijs committed:
Change DisposableEventListener to be just a function that returns the dispose method

Bump major version number since this is a breaking change

mediamonks-thijs committed:
Rename DisposableEventListener to addEventListener since it's no longer an object, but a function

ReneDrie commented:
I do like the naming (addEventListener) is shorter and makes sense. I don't like the way of disposing.
Seeing `windowResizeListener()` isn't very declarative in saying that it's clearing the listener. I'd rather see listener.dispose() then.

ThaNarie commented:
@ReneDrie  Then name it `windowResizeListenerDisposer()` - all up to you in user-land :)

ReneDrie commented:
Haha yeah that's what I thought as well. Ok seems fine to me then

mediamonks-thijs committed:
Update example to latest changes

mediamonks-thijs committed:
Update README.md

mediamonks-thijs committed:
update tslint

mediamonks-thijs committed:
update README.md

mediamonks-thijs committed:
Update typing to support events that extend Event, like KeyboardEvent
Add unit test to test KeyboardEvents
update Example to add KeyboardEvent handling
